### PR TITLE
it fixes make wallet file writes atomic to prevent data loss on crash

### PR DESCRIPTION
**Pull request by LusterSourav**
## Problem

write_to_disk in storage.rs uses truncate(true) to open the wallet file. This empties the file before new data is written. If the process crashes during the write window, the wallet file is left at 0 bytes. load_sensitive_struct panics on empty files with no recovery path, locking users out of their wallets permanently.

This is the same pattern that affected Bitcoin Core v30 (wallet file deletion), Electrum (551+ corruption reports), and Sparrow (unrecoverable loss).

## Fix

write_to_disk now writes to a .cbor.tmp file then renames over the original. This matches the pattern already used by MakerSwapTracker::flush (swap_tracker.rs:196-212) and TakerSwapTracker::flush (taker/swap_tracker.rs:448-462) in this same codebase.

Also removes the redundant File::create call in init (line 104), since write_to_disk now handles file creation via the tmp file.

## Changes

- src/wallet/storage.rs: 1 file, 41 insertions, 12 deletions
- Added test_atomic_write_never_leaves_empty_file

## Verification

- All 81 unit tests pass
- Commit is GPG signed

fsync before rename and load-time validation for corrupt files. These are tracked as follow-ups in #926.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Protocol Version(s) Affected

- [x] Neither (infrastructure / tooling only)

## Affected Component(s)

- [x] Core protocol / cryptography / library

### Code Quality
- [x] Pre-commit git hook passes

### Testing
- [x] All unit tests pass (`cargo test`)

### Security & Privacy
- [x] Edge cases and error handling considered

## How to Test

```bash
cargo test -p coinswap --lib wallet::storage::tests -- --nocapture

---


<!-- This is an auto-generated comment: release notes by coderabbit.ai —>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved wallet saving to reduce the risk of empty or partially written wallet files after an interruption.
  * Wallet data is now written more safely during updates, helping preserve stored information more reliably.
  * Added checks to confirm temporary files are cleaned up properly after wallet writes.

<!-- end of auto-generated comment: release notes by coderabbit.ai —>

### CODE REVIEW

**Comment by LusterSourav** [comment]
Wallet file was being emptied before rewrite
so Crash = 0-byte file = permanent lockout. Now writes to tmp then renames. Original never touched until new data is confirmed on disk as load_sensitive_struct panics on empty files. No retry, no recovery

Wallet write is done for now . Two more files have the same bug need some time to write 
# to be noted ...

File: src/wallet/storage.rs [hunk 1]
```diff
@@ -118,27 +115,26 @@ impl WalletStore {
...
-                // Write encrypted wallet data to disk.
                 serde_cbor::to_writer(writer, &encrypted)?;
             }
             None => {
-                // No encryption: serialize and write the wallet directly.
                 serde_cbor::to_writer(writer, &self)?;
             }
         }
+
+        fs::rename(&tmp_path, path)?;
```
[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:137]: the swap trackers{ (maker/swap_tracker.rs:196-212, taker/swap_tracker.rs:448-462) }don't use fsync either. They just do write + rename. This PR matches the existing codebase pattern. adding fsync is a valid durability improvement but it's scope creep for this pr. It's a separate issue Lol..

[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:137]: so # skip hmmm

File: src/wallet/storage.rs [hunk 2]
```diff
@@ -118,27 +115,26 @@ impl WalletStore {
...
             .truncate(true)
-            .open(path)?;
+            .open(&tmp_path)?;
         let writer = BufWriter::new(wallet_file);
 
         match store_enc_material {
             Some(material) => {
-                // Encryption branch: encrypt the serialized wallet before writing.
-
                 let encrypted = encrypt_struct(self, material).unwrap();
```
[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:129]: wt the .unwrap() on encrypt_struct was pre-existing code. i just moved it. my pr is about atomic writes, not error handling cleanup. Adding ? would change the error handling contract of write_to_disk — that's a different change.

File: src/wallet/storage.rs [hunk 3]
```diff
@@ -189,4 +185,37 @@ mod tests {
...
+        let bytes = fs::read(&file_path).unwrap();
+        assert!(!bytes.is_empty(), "wallet file should not be empty after init");
+        assert!(!tmp_path.exists(), "tmp file should not exist after rename");
+
+        // Write again: tmp file is created then renamed over original
+        store.write_to_disk(&file_path, &None).unwrap();
+        let bytes = fs::read(&file_path).unwrap();
+        assert!(!bytes.is_empty(), "wallet file should not be empty after second write");
+        assert!(!tmp_path.exists(), "tmp file should not exist after rename");
+    }
```
[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:220]: us test proves the core claim file is never empty between writes. Simulating a crash in a unit test requires mocking the filesystem (kill process mid-write). That's integration test territory, not unit test. The test does what it was designed to do oke doki my rabit

[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:220]: so skippp

File: src/wallet/storage.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LusterSourav) [reply src/wallet/storage.rs:129]: so skipppss

**Comment by LusterSourav** [comment]
hohohho 
i like this poem
A rabbit hops with cautious cheer,
No half-saved wallets far or near,
Temp file first, then swap in place,
A crash can't wipe a single trace,
Hop hop, my coins are safe and sound! 🐇💰
